### ISSUE TITLE
Bazel test flake: //test/cpp/end2end:client_lb_end2end_test is flaky

### ISSUE THREAD

**Issue by adelez** [grpc/grpc#16123]
https://source.cloud.google.com/results/invocations/cb65cb0d-e989-47e9-8534-aa9ea1b5c12e/targets/%2F%2Ftest%2Fcpp%2Fend2end:client_lb_end2end_test@poller%3Depoll1/tests

**Comment by jtattermusch** [comment]
This is currently the top flake on remote bazel (it fails 40% of builds).

Happening for all of: tsan, asan, opt
Pollers: poll, epoll1, poll-cv, epollsig

Most common occurence is on tsan. Recent failures:
https://source.cloud.google.com/results/invocations/1e51fa49-6c97-4b12-adab-a54f85e0f406
https://source.cloud.google.com/results/invocations/b99d03e9-72f9-4dea-8d5f-3da693e66998
https://source.cloud.google.com/results/invocations/e597353b-08de-4b0d-9bdd-138d490807b5
https://source.cloud.google.com/results/invocations/03fd2625-8505-497d-a53c-44307afd8884
https://source.cloud.google.com/results/invocations/fc5fea92-3e8a-474b-bd6e-62d0c2b4e22e

**Comment by jtattermusch** [comment]
@yashykt  can you take a look?

**Comment by yashykt** [comment]
These are usually failing with Deadline Exceeded or Connect Failed. Seeing that this is only happening for bazel, this is highly likely related to #15610 and #14240

**Comment by jtattermusch** [comment]
The assertions in the test are:
https://github.com/grpc/grpc/blob/2548704bee6dd9071832135f8a76c2761897cddb/test/cpp/end2end/client_lb_end2end_test.cc#L221

https://github.com/grpc/grpc/blob/2548704bee6dd9071832135f8a76c2761897cddb/test/cpp/end2end/client_lb_end2end_test.cc#L566

**Comment by jtattermusch** [comment]
We might have better data once https://github.com/grpc/grpc/pull/16478 is merged.

**Comment by jtattermusch** [comment]
After enabling GRPC_VERBOSITY=debug  on foundry tests, here is the more detailed log:

**Comment by jtattermusch** [comment]
What's odd in the more detailed logs is that the PickFirstManyUpdates test case  start servers on ports 57515, 57516 and 57517,
but the connection failure that keeps happening is for target `ipv4:127.0.0.1:57507` (= different port).
There's also a bunch of `starting server on port 57xxx` messages throughout the test suite, but the one 
that says `starting server on port 57507` is missing (but 57505, 57506  and 57508 are there, just 57507 is missing).

@dgquintas  can you please doublecheck that such behavior is expected?


```
[ RUN      ] ClientLbEnd2endTest.PickFirstManyUpdates
I0829 03:26:40.744683669      19 client_lb_end2end_test.cc:259] starting server on port 57515
I0829 03:26:40.744782556      77 server_builder.cc:270]      Synchronous server. Num CQs: 1, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10000
I0829 03:26:40.745080355      19 client_lb_end2end_test.cc:267] server startup complete
I0829 03:26:40.745095133      19 client_lb_end2end_test.cc:259] starting server on port 57516
I0829 03:26:40.745216167      79 server_builder.cc:270]      Synchronous server. Num CQs: 1, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10000
I0829 03:26:40.745549894      19 client_lb_end2end_test.cc:267] server startup complete
I0829 03:26:40.745579287      19 client_lb_end2end_test.cc:259] starting server on port 57517
I0829 03:26:40.745722646      81 server_builder.cc:270]      Synchronous server. Num CQs: 1, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10000
I0829 03:26:40.746056696      19 client_lb_end2end_test.cc:267] server startup complete
I0829 03:26:40.746149593      19 client_lb_end2end_test.cc:575] Force subchannel creation: 1
I0829 03:26:46.124809250      22 subchannel.cc:668]          Connect failed: {"created":"@1535513206.124692187","description":"Failed to connect to remote host: FD shutdown","file":"src/core/lib/iomgr/ev_poll_posix.cc","file_line":532,"grpc_status":14,"os_error":"Timeout occurred","referenced_errors":[{"created":"@1535513206.124689800","description":"connect() timed out","file":"src/core/lib/iomgr/tcp_client_posix.cc","file_line":114}],"target_address":"ipv4:127.0.0.1:57507"}
I0829 03:26:46.124888313      19 subchannel.cc:490]          Subchannel 0xed9760: Retry immediately
I0829 03:26:46.124907968      19 subchannel.cc:452]          Failed to connect to channel, retrying
I0829 03:26:46.125408450      19 subchannel.cc:668]          Connect failed: {"created":"@1535513206.125344845","description":"Failed to connect to remote host: OS Error","errno":111,"file":"src/core/lib/iomgr/tcp_client_posix.cc","file_line":202,"os_error":"Connection refused","syscall":"connect","target_address":"ipv4:127.0.0.1:57507"}
I0829 03:26:46.125440568      19 subchannel.cc:492]          Subchannel 0xed9760: Retry in 4902 milliseconds
test/cpp/end2end/client_lb_end2end_test.cc:232: Failure
```

**Comment by yashykt** [comment]
does that mean that the server with port 57507 is not being started?

**Comment by dgquintas** [comment]
Of course, this doesn't repro locally... ran 1000 times with bazel locally. I'm trying to address some issues I'm having with executing this remotely, but yes, I'm having a look.

It's indeed very strange that it's trying to connect to that port, which indeed doesn't correspond to any server. [The test code is very straightforward, ](https://github.com/grpc/grpc/blob/5ee9b5e7b547d0a82c84cc3767236ace8d5cd469/test/cpp/end2end/client_lb_end2end_test.cc#L563)there's no room for a trivial error in there. @AspirinSJL can you think of any way in which we could be modifying the port in pick first?

**Comment by dgquintas** [comment]
A possible scenario I can't currently verify because I can't execute tests remotely:

ServerData instances are too slow to be created [here](https://github.com/grpc/grpc/blob/5ee9b5e7b547d0a82c84cc3767236ace8d5cd469/test/cpp/end2end/client_lb_end2end_test.cc#L138), and the test then reads stale (or racy) values [here](https://github.com/grpc/grpc/blob/5ee9b5e7b547d0a82c84cc3767236ace8d5cd469/test/cpp/end2end/client_lb_end2end_test.cc#L571). That is, because there's no synchronization between lines 566 (where the writes to server->port happen) and 571 (where port values are read).

**Comment by jtattermusch** [comment]
@dgquintas  looks like both happen on the same (main) thread so no synchronization should be needed?

**Comment by dgquintas** [comment]
Something I'm seeing in all the local failures (where tests are still run remotely) is:
```
I0830 22:04:40.035325108      17 client_lb_end2end_test.cc:259] starting server on port 53740
I0830 22:04:40.035567980     145 chttp2_server.cc:320]       WARNING: {"created":"@1535666680.035539573","description":"Only 1 addresses added out of total 2 resolved","file":"src/core/ext/transport/chttp2/server/chttp2_server.cc","file_line":316,"referenced_errors":[{"created":"@1535666680.035511738","description":"Unable to configure socket","fd":12,"file":"src/core/lib/iomgr/tcp_server_utils_posix_common.cc","file_line":214,"referenced_errors":[{"created":"@1535666680.035504501","description":"OS Error","errno":98,"file":"src/core/lib/iomgr/tcp_server_utils_posix_common.cc","file_line":187,"os_error":"Address already in use","syscall":"bind"}]}]}
```
In that case, we see that the test is trying to fire up a server at port `53740` but the address was already in use, so no server is actually started. Following client lines are all like

```I0830 22:04:45.004143549      28 subchannel.cc:668]          Connect failed: {"created":"@1535666685.004105858","description":"Failed to connect to remote host: OS Error","errno":111,"file":"src/core/lib/iomgr/tcp_client_posix.cc","file_line":202,"os_error":"Connection refused","syscall":"connect","target_address":"ipv4:127.0.0.1:53740"}```

confirming that the failure is caused by the misfire of the server startup.

Now, it seems to me the port server the tests rely on is no acting correctly. Perhaps more than one instance is running, or the OS is not reclaiming ports quickly enough.

**Comment by dgquintas** [comment]
I don't think we want to tweak the tests to handle failures to start servers (it's not clear to me how we'd do that... that warning doesn't propagate high enough I think): the whole point of the port server is to guarantee the ports are available.

**Comment by adelez** [comment]
So essentially it's a dup of #14240. I'm ok with closing this bug if that's the case.

**Comment by jtattermusch** [comment]
@dgquintas  I don't see the failure you mentioned in any of the runs mentioned here: https://github.com/grpc/grpc/issues/16123#issuecomment-416834261  (and in my manually spawned runs, which actually seem to be failing with much lower rate than on Kokoro and not always in PickFirstManyUpdates).

Why I don't think this is just a dup of #14240 (it's most likely a contributing factor, but not the full cause):
- on Foundry, this test almost always fails in PickFirstManyUpdates - that would be a very strange coincidence if this was #14240 (which should be roughly equally distributed).
- client_lb_end2end_test is several times more flaky than any other tests that are also affected by #14240. Probably not a coincidence.

Until this is explained, this issue should stay open.

**Comment by jtattermusch** [comment]
Based on historical data it looks like vast majority of occurrences is happening for `GRPC_POLL_STRATEGY=poll-cv` (i.e. for `//test/cpp/end2end:client_lb_end2end_test@poller=poll-cv`)

**Comment by jtattermusch** [comment]
Data seem to indicate that this was fixed by #16620.
